File: ios/Cove/SettingsScreen/DecoyPinView.swift
//
//  DecoyPinView.swift
//  Cove
//
//  Created by Praveen Perera on 01/28/24.
//

import SwiftUI

private enum PinState {
    case new
    case confirm(String)
}

struct DecoyPinView: View {
    /// args
    var onComplete: (String) -> Void
    var backAction: () -> Void

    /// private
    @State private var pinState: PinState = .new

    var body: some View {
        Group {
            switch pinState {
            case .new:
                NumberPadPinView(
                    title: "Enter Decoy PIN",
                    isPinCorrect: { _ in true },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: { enteredPin in
                        withAnimation {
                            pinState = .confirm(enteredPin)
                        }
                    }
                )
            case let .confirm(pinToConfirm):
                NumberPadPinView(
                    title: "Confirm Decoy PIN",
                    isPinCorrect: { $0 == pinToConfirm },
                    showPin: false,
                    backAction: backAction,
                    onUnlock: onComplete
                )
            }
        }
    }
}

#Preview {
    DecoyPinView(onComplete: { _ in }, backAction: {})
        .environment(AuthManager.shared)
}

File: android/app/src/main/java/org/bitcoinppl/cove/settings/DecoyPinView.kt
package org.bitcoinppl.cove.settings

import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import org.bitcoinppl.cove.views.NumberPadPinView

private sealed class DecoyPinState {
    data object New : DecoyPinState()

    data class Confirm(
        val pinToConfirm: String,
    ) : DecoyPinState()
}

@Composable
fun DecoyPinView(
    onComplete: (String) -> Unit,
    backAction: () -> Unit,
) {
    var pinState: DecoyPinState by remember { mutableStateOf(DecoyPinState.New) }

    when (val state = pinState) {
        is DecoyPinState.New -> {
            NumberPadPinView(
                title = "Enter Decoy PIN",
                isPinCorrect = { true },
                showPin = false,
                backAction = backAction,
                onUnlock = { enteredPin ->
                    pinState = DecoyPinState.Confirm(enteredPin)
                },
            )
        }

        is DecoyPinState.Confirm -> {
            NumberPadPinView(
                title = "Confirm Decoy PIN",
                isPinCorrect = { it == state.pinToConfirm },
                showPin = false,
                backAction = backAction,
                onUnlock = onComplete,
            )
        }
    }
}
